lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -478,185 +491,226 @@ class PullbackInfo {
+      linearMapStruct->print(s);
       s << '\n';
     });
-    return pullbackStruct;
+    return linearMapStruct;
   }
 
 public:
-  PullbackInfo(const PullbackInfo &) = delete;
-  PullbackInfo &operator=(const PullbackInfo &) = delete;
+  LinearMapInfo(const LinearMapInfo &) = delete;
+  LinearMapInfo &operator=(const LinearMapInfo &) = delete;
 
-  explicit PullbackInfo(ADContext &context, SILFunction *original,
-                        SILFunction *vjp, const SILAutoDiffIndices &indices);
+  explicit LinearMapInfo(ADContext &context,
+                         AutoDiffAssociatedFunctionKind kind,
+                         SILFunction *original, SILFunction *assocFn,
+                         const SILAutoDiffIndices &indices);
 
-  /// Returns the pullback struct associated with the given original block.
-  StructDecl *getPullbackStruct(SILBasicBlock *origBB) const {
-    return pullbackStructs.lookup(origBB);
+  /// Returns the linear map struct associated with the given original block.
+  StructDecl *getLinearMapStruct(SILBasicBlock *origBB) const {
+    return linearMapStructs.lookup(origBB);
   }
 
-  /// Returns the lowered SIL type of the pullback struct associated with the
+  /// Returns the lowered SIL type of the linear map struct associated with the
   /// given original block.
-  SILType getPullbackStructLoweredType(SILBasicBlock *origBB) const {
-    auto *pbStruct = getPullbackStruct(origBB);
-    auto pbStructType =
-        pbStruct->getDeclaredInterfaceType()->getCanonicalType();
-    return typeConverter.getLoweredType(pbStructType,
+  SILType getLinearMapStructLoweredType(SILBasicBlock *origBB) const {
+    auto *linMapStruct = getLinearMapStruct(origBB);
+    auto linMapStructType =
+        linMapStruct->getDeclaredInterfaceType()->getCanonicalType();
+    return typeConverter.getLoweredType(linMapStructType,
                                         ResilienceExpansion::Minimal);
   }
 
-  /// Returns the predecessor enum associated with the given original block.
-  EnumDecl *getPredecessorEnum(SILBasicBlock *origBB) const {
-    return predecessorEnums.lookup(origBB);
+  /// Returns the branching trace enum associated with the given original block.
+  EnumDecl *getBranchingTraceDecl(SILBasicBlock *origBB) const {
+    return branchingTraceDecls.lookup(origBB);
   }
 
-  /// Returns the lowered SIL type of the predecessor enum associated with the
-  /// given original block.
-  SILType getPredecessorEnumLoweredType(SILBasicBlock *origBB) const {
-    auto *predEnum = getPredecessorEnum(origBB);
-    auto predEnumType =
-        predEnum->getDeclaredInterfaceType()->getCanonicalType();
-    return typeConverter.getLoweredType(predEnumType,
+  /// Returns the lowered SIL type of the branching trace enum associated with
+  /// the given original block.
+  SILType getBranchingTraceEnumLoweredType(SILBasicBlock *origBB) const {
+    auto *linMapEnum = getBranchingTraceDecl(origBB);
```

Comment:
```suggestion
    auto *traceEnum = getBranchingTraceDecl(origBB);
```

Reply:
Will call it `trace` as suggested in chat

Reply:
`traceDecl` would be better because this is not a trace value.